File: _Basic/windows-forms-introduction-with-visual-basic/7-windows-forms-introduction-with-visual-basic-m7-datasets-exercise-files/VB/Data/Data/DataForm.vb
Public Class DataForm
    Private _categoriesBindingSource As New BindingSource()
    Private _productsBindingSource As New BindingSource()

    Private _currentSource As ISource
    Private _objectSource As ObjectSource
    Private _dataSetSource As DataSetSource

    Private Sub DataForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        SourceToolStripComboBox.SelectedIndex = 0
        SetSource()
        BindCategories()

        _categoriesBindingSource.DataSource = _currentSource.GetCategories()

        CategoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName"
        CategoryToolStripComboBox.ComboBox.ValueMember = "CategoryID"
        CategoryToolStripComboBox.ComboBox.DataSource = _categoriesBindingSource

        ProductsDataGridView.DataSource = _productsBindingSource
        ProductsListBox.DisplayMember = "ProductName"
        ProductsListBox.ValueMember = "ProductID"
        ProductsListBox.DataSource = _productsBindingSource

        NameTextBox.DataBindings.Add("Text", _productsBindingSource, "ProductName")
        QuantityTextBox.DataBindings.Add("Text", _productsBindingSource, "QuantityPerUnit")
        PriceTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitPrice")
        StockTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsInStock")
        OrderTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsOnOrder")
        DiscontinuedCheckBox.DataBindings.Add("Checked", _productsBindingSource, "Discontinued")
    End Sub

    Private Sub SourceToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles SourceToolStripComboBox.SelectedIndexChanged
        SetSource()
        BindCategories()
        BindProducts()
    End Sub

    Private Sub CategoryToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CategoryToolStripComboBox.SelectedIndexChanged
        BindProducts()
    End Sub

    Private Sub SetSource()
        Select Case SourceToolStripComboBox.SelectedIndex
            Case 0
                If _objectSource Is Nothing Then
                    _objectSource = New ObjectSource()
                End If
                _currentSource = _objectSource
            Case 1
                If _dataSetSource Is Nothing Then
                    _dataSetSource = New DataSetSource()
                End If
                _currentSource = _dataSetSource
        End Select
    End Sub

    Private Sub BindCategories()
        _categoriesBindingSource.DataSource = _currentSource.GetCategories()
    End Sub

    Private Sub BindProducts()
        Dim catId = CategoryToolStripComboBox.ComboBox.SelectedValue
        _productsBindingSource.DataSource = _currentSource.GetProducts(catId)
    End Sub

    Private Sub AddToolStripButton_Click(sender As Object, e As EventArgs) Handles AddToolStripButton.Click
        Dim id = CategoryToolStripComboBox.ComboBox.SelectedValue
        Dim name = CategoryToolStripComboBox.Text
        Dim category As New Category(id, name)

        Dim form As New AddProductForm(category)
        Dim result = form.ShowDialog()

        If result = Windows.Forms.DialogResult.OK Then
            _currentSource.AddProduct(_productsBindingSource, form.Product)
        End If
    End Sub

    Private Sub DeleteToolStripButton_Click(sender As Object, e As EventArgs) Handles DeleteToolStripButton.Click
        Dim productId = ProductsListBox.SelectedValue
        _currentSource.DeleteProduct(_productsBindingSource, productId)
    End Sub

    Private Sub SaveToolStripButton_Click(sender As Object, e As EventArgs) Handles SaveToolStripButton.Click
        _currentSource.Save()
    End Sub

    Private Sub BackToolStripButton_Click(sender As Object, e As EventArgs) Handles BackToolStripButton.Click
        _productsBindingSource.MovePrevious()
    End Sub

    Private Sub ForwardToolStripButton_Click(sender As Object, e As EventArgs) Handles ForwardToolStripButton.Click
        _productsBindingSource.MoveNext()
    End Sub

End Class

File: _Basic/windows-forms-introduction-with-visual-basic/7-windows-forms-introduction-with-visual-basic-m7-datasets-exercise-files/CS/Data/Data/DataForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Data
{
    public partial class DataForm : Form
    {
        public DataForm()
        {
            InitializeComponent();
        }

        private BindingSource _categoriesBindingSource = new BindingSource();
        private BindingSource _productsBindingSource = new BindingSource();

        private ISource _currentSource;
        private ObjectSource _objectSource;
        private DataSetSource _dataSetSource;

        private void DataForm_Load(object sender, EventArgs e)
        {
            try
            {
                SourceToolStripComboBox.SelectedIndex = 0;
                SetSource();
                BindCategories();

                _categoriesBindingSource.DataSource = _currentSource.GetCategories();

                CategoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName";
                CategoryToolStripComboBox.ComboBox.ValueMember = "CategoryID";
                CategoryToolStripComboBox.ComboBox.DataSource = _categoriesBindingSource;

                ProductsDataGridView.DataSource = _productsBindingSource;
                ProductsListBox.DisplayMember = "ProductName";
                ProductsListBox.ValueMember = "ProductID";
                ProductsListBox.DataSource = _productsBindingSource;

                NameTextBox.DataBindings.Add("Text", _productsBindingSource, "ProductName");
                QuantityTextBox.DataBindings.Add("Text", _productsBindingSource, "QuantityPerUnit");
                PriceTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitPrice");
                StockTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsInStock");
                OrderTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsOnOrder");
                DiscontinuedCheckBox.DataBindings.Add("Checked", _productsBindingSource, "Discontinued");
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }

        private void SourceToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            SetSource();
            BindCategories();
            BindProducts();
        }

        private void CategoryToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            BindProducts();
        }

        private void SetSource()
        {
            switch (SourceToolStripComboBox.SelectedIndex)
            {
                case 0:
                    if (_objectSource == null)
                    {
                        _objectSource = new ObjectSource();
                    }
                    _currentSource = _objectSource;
                    break;
                case 1:
                    if (_dataSetSource == null)
                    {
                        _dataSetSource = new DataSetSource();
                    }
                    _currentSource = _dataSetSource;
                    break;
            }
        }

        private void BindCategories()
        {
            _categoriesBindingSource.DataSource = _currentSource.GetCategories();
        }

        private void BindProducts()
        {
            var catId = Convert.ToInt32(CategoryToolStripComboBox.ComboBox.SelectedValue);
            _productsBindingSource.DataSource = _currentSource.GetProducts(catId);
        }

        private void AddToolStripButton_Click(object sender, EventArgs e)
        {
            var id = Convert.ToInt32(CategoryToolStripComboBox.ComboBox.SelectedValue);
            var name = CategoryToolStripComboBox.Text;
            var category = new Category(id, name);

            AddProductForm form = new AddProductForm(category);
            var result = form.ShowDialog();

            if (result == DialogResult.OK)
            {
                _currentSource.AddProduct(_productsBindingSource, form.Product);
            }
        }

        private void DeleteToolStripButton_Click(object sender, EventArgs e)
        {
            var productId = Convert.ToInt32(ProductsListBox.SelectedValue);
            _currentSource.DeleteProduct(_productsBindingSource, productId);
        }

        private void SaveToolStripButton_Click(object sender, EventArgs e)
        {
            _currentSource.Save();
        }

        private void BackToolStripButton_Click(object sender, EventArgs e)
        {
            _productsBindingSource.MovePrevious();
        }

        private void ForwardToolStripButton_Click(object sender, EventArgs e)
        {
            _productsBindingSource.MoveNext();
        }
    }
}
